// powerbi-api/lib/autorest/models/basicCredentials.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * @class
 * Initializes a new instance of the BasicCredentials class.
 * @constructor
 * Object representing basic authentication credentials
 * @member {string} [username] Username required to access the datasource
 * 
 * @member {string} [password] Password required to access the datasource
 * 
 */
function BasicCredentials() {
}

/**
 * Defines the metadata of BasicCredentials
 *
 * @returns {object} metadata of BasicCredentials
 *
 */
BasicCredentials.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'BasicCredentials',
    type: {
      name: 'Composite',
      className: 'BasicCredentials',
      modelProperties: {
        username: {
          required: false,
          serializedName: 'username',
          type: {
            name: 'String'
          }
        },
        password: {
          required: false,
          serializedName: 'password',
          type: {
            name: 'String'
          }
        }
      }
    }
  };
};

module.exports = BasicCredentials;
