// powerbi-api/lib/autorest/models/column.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * @class
 * Initializes a new instance of the Column class.
 * @constructor
 * A Power BI Column
 * @member {string} name The column name
 * 
 * @member {string} dataType The column data type
 * 
 */
function Column() {
}

/**
 * Defines the metadata of Column
 *
 * @returns {object} metadata of Column
 *
 */
Column.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'Column',
    type: {
      name: 'Composite',
      className: 'Column',
      modelProperties: {
        name: {
          required: true,
          serializedName: 'name',
          constraints: {
            Pattern: '^[\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u10000-\u10FFFF]+$'
          },
          type: {
            name: 'String'
          }
        },
        dataType: {
          required: true,
          serializedName: 'dataType',
          type: {
            name: 'String'
          }
        }
      }
    }
  };
};

module.exports = Column;
