// powerbi-api/lib/autorest/models/createGatewayRequest.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

var models = require('./index');

var util = require('util');

/**
 * @class
 * Initializes a new instance of the CreateGatewayRequest class.
 * @constructor
 * The gateway info
 * @member {string} [name] The gateway name
 * 
 * @member {string} [publicKey] The gateway public key
 * 
 * @member {string} [annotation] The gateway annotation
 * 
 * @member {array} [workspaces] The workspace id to add the gateway to
 * 
 */
function CreateGatewayRequest() {
}

/**
 * Defines the metadata of CreateGatewayRequest
 *
 * @returns {object} metadata of CreateGatewayRequest
 *
 */
CreateGatewayRequest.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'CreateGatewayRequest',
    type: {
      name: 'Composite',
      className: 'CreateGatewayRequest',
      modelProperties: {
        name: {
          required: false,
          serializedName: 'name',
          type: {
            name: 'String'
          }
        },
        publicKey: {
          required: false,
          serializedName: 'publicKey',
          type: {
            name: 'String'
          }
        },
        annotation: {
          required: false,
          serializedName: 'annotation',
          type: {
            name: 'String'
          }
        },
        workspaces: {
          required: false,
          serializedName: 'workspaces',
          type: {
            name: 'Sequence',
            element: {
                required: false,
                serializedName: 'WorkspaceIdElementType',
                type: {
                  name: 'Composite',
                  className: 'WorkspaceId'
                }
            }
          }
        }
      }
    }
  };
};

module.exports = CreateGatewayRequest;
