// powerbi-api/lib/autorest/models/datasource.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * @class
 * Initializes a new instance of the Datasource class.
 * @constructor
 * A Power BI datasource
 * @member {string} [name] The datasource name
 * 
 * @member {string} [connectionString] The datasource connection string
 * 
 */
function Datasource() {
}

/**
 * Defines the metadata of Datasource
 *
 * @returns {object} metadata of Datasource
 *
 */
Datasource.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'Datasource',
    type: {
      name: 'Composite',
      className: 'Datasource',
      modelProperties: {
        name: {
          required: false,
          serializedName: 'name',
          type: {
            name: 'String'
          }
        },
        connectionString: {
          required: false,
          serializedName: 'connectionString',
          type: {
            name: 'String'
          }
        }
      }
    }
  };
};

module.exports = Datasource;
