// powerbi-api/lib/autorest/models/dataset.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

var models = require('./index');

var util = require('util');

/**
 * @class
 * Initializes a new instance of the Dataset class.
 * @constructor
 * A Power BI Dataset
 * @member {string} [id] The dataset id
 * 
 * @member {string} name The dataset name
 * 
 * @member {string} [defaultRetentionPolicy] The dataset default data
 * retention policy
 * 
 * @member {array} tables The dataset tables
 * 
 * @member {string} [webUrl] The dataset web url
 * 
 * @member {array} [datasources] The datasources associated with this dataset
 * 
 */
function Dataset() {
}

/**
 * Defines the metadata of Dataset
 *
 * @returns {object} metadata of Dataset
 *
 */
Dataset.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'Dataset',
    type: {
      name: 'Composite',
      className: 'Dataset',
      modelProperties: {
        id: {
          required: false,
          serializedName: 'id',
          type: {
            name: 'String'
          }
        },
        name: {
          required: true,
          serializedName: 'name',
          type: {
            name: 'String'
          }
        },
        defaultRetentionPolicy: {
          required: false,
          serializedName: 'defaultRetentionPolicy',
          type: {
            name: 'String'
          }
        },
        tables: {
          required: true,
          serializedName: 'tables',
          type: {
            name: 'Sequence',
            element: {
                required: false,
                serializedName: 'TableElementType',
                type: {
                  name: 'Composite',
                  className: 'Table'
                }
            }
          }
        },
        webUrl: {
          required: false,
          serializedName: 'webUrl',
          type: {
            name: 'String'
          }
        },
        datasources: {
          required: false,
          serializedName: 'datasources',
          type: {
            name: 'Sequence',
            element: {
                required: false,
                serializedName: 'DatasourceElementType',
                type: {
                  name: 'Composite',
                  className: 'Datasource'
                }
            }
          }
        }
      }
    }
  };
};

module.exports = Dataset;
