// powerbi-api/lib/autorest/models/gatewayDatasource.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

var models = require('./index');

/**
 * @class
 * Initializes a new instance of the GatewayDatasource class.
 * @constructor
 * A Power BI gateway datasource
 * @member {string} [id] The unique id for this gateway datasource
 * 
 * @member {string} [gatewayId] The associated gateway id
 * 
 * @member {string} [datasourceType] The datasource type
 * 
 * @member {string} [connectionDetails] The datasource connection details
 * 
 * @member {string} [credentialType] The datasource credential type
 * 
 * @member {object} [basicCredentials]
 * 
 * @member {string} [basicCredentials.username] Username required to access
 * the datasource
 * 
 * @member {string} [basicCredentials.password] Password required to access
 * the datasource
 * 
 */
function GatewayDatasource() {
}

/**
 * Defines the metadata of GatewayDatasource
 *
 * @returns {object} metadata of GatewayDatasource
 *
 */
GatewayDatasource.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'GatewayDatasource',
    type: {
      name: 'Composite',
      className: 'GatewayDatasource',
      modelProperties: {
        id: {
          required: false,
          serializedName: 'id',
          type: {
            name: 'String'
          }
        },
        gatewayId: {
          required: false,
          serializedName: 'gatewayId',
          type: {
            name: 'String'
          }
        },
        datasourceType: {
          required: false,
          serializedName: 'datasourceType',
          type: {
            name: 'String'
          }
        },
        connectionDetails: {
          required: false,
          serializedName: 'connectionDetails',
          type: {
            name: 'String'
          }
        },
        credentialType: {
          required: false,
          serializedName: 'credentialType',
          type: {
            name: 'String'
          }
        },
        basicCredentials: {
          required: false,
          serializedName: 'basicCredentials',
          type: {
            name: 'Composite',
            className: 'BasicCredentials'
          }
        }
      }
    }
  };
};

module.exports = GatewayDatasource;
