// powerbi-api/lib/autorest/models/gatewayPublicKey.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * @class
 * Initializes a new instance of the GatewayPublicKey class.
 * @constructor
 * A Power BI Gateway Public Key
 * @member {string} [exponent] The public key exponent
 * 
 * @member {string} [modulus] The public key  modulus
 * 
 */
function GatewayPublicKey() {
}

/**
 * Defines the metadata of GatewayPublicKey
 *
 * @returns {object} metadata of GatewayPublicKey
 *
 */
GatewayPublicKey.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'GatewayPublicKey',
    type: {
      name: 'Composite',
      className: 'GatewayPublicKey',
      modelProperties: {
        exponent: {
          required: false,
          serializedName: 'exponent',
          type: {
            name: 'String'
          }
        },
        modulus: {
          required: false,
          serializedName: 'modulus',
          type: {
            name: 'String'
          }
        }
      }
    }
  };
};

module.exports = GatewayPublicKey;
