// powerbi-api/lib/autorest/models/index.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

/* jshint latedef:false */
/* jshint forin:false */
/* jshint noempty:false */

'use strict';

exports.ImportModel = require('./importModel');
exports.Report = require('./report');
exports.Dataset = require('./dataset');
exports.Table = require('./table');
exports.Column = require('./column');
exports.Row = require('./row');
exports.Datasource = require('./datasource');
exports.ImportInfo = require('./importInfo');
exports.ODataResponseListDataset = require('./oDataResponseListDataset');
exports.ODataResponseListTable = require('./oDataResponseListTable');
exports.ODataResponseListGatewayDatasource = require('./oDataResponseListGatewayDatasource');
exports.GatewayDatasource = require('./gatewayDatasource');
exports.BasicCredentials = require('./basicCredentials');
exports.ODataResponseListDatasource = require('./oDataResponseListDatasource');
exports.ODataResponseListReport = require('./oDataResponseListReport');
exports.ODataResponseListImport = require('./oDataResponseListImport');
exports.ODataResponseListWorkspace = require('./oDataResponseListWorkspace');
exports.Workspace = require('./workspace');
exports.ODataResponseListGateway = require('./oDataResponseListGateway');
exports.Gateway = require('./gateway');
exports.GatewayPublicKey = require('./gatewayPublicKey');
exports.WorkspaceId = require('./workspaceId');
exports.ODataResponseString = require('./oDataResponseString');
exports.CreateGatewayRequest = require('./createGatewayRequest');
