// powerbi-api/lib/autorest/models/oDataResponseString.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * @class
 * Initializes a new instance of the ODataResponseString class.
 * @constructor
 * Odata response wrapper for a string
 * @member {string} [odatacontext]
 * 
 * @member {string} [value] The returned value
 * 
 */
function ODataResponseString() {
}

/**
 * Defines the metadata of ODataResponseString
 *
 * @returns {object} metadata of ODataResponseString
 *
 */
ODataResponseString.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'ODataResponse[String]',
    type: {
      name: 'Composite',
      className: 'ODataResponseString',
      modelProperties: {
        odatacontext: {
          required: false,
          serializedName: 'odata\\.context',
          type: {
            name: 'String'
          }
        },
        value: {
          required: false,
          serializedName: 'value',
          type: {
            name: 'String'
          }
        }
      }
    }
  };
};

module.exports = ODataResponseString;
