// powerbi-api/lib/autorest/models/oDataResponseListTable.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

var models = require('./index');

var util = require('util');

/**
 * @class
 * Initializes a new instance of the ODataResponseListTable class.
 * @constructor
 * Odata response wrapper for a Power BI Table collection
 * @member {string} [odatacontext]
 * 
 * @member {array} [value] The Power BI tables
 * 
 */
function ODataResponseListTable() {
}

/**
 * Defines the metadata of ODataResponseListTable
 *
 * @returns {object} metadata of ODataResponseListTable
 *
 */
ODataResponseListTable.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'ODataResponse[List[Table]]',
    type: {
      name: 'Composite',
      className: 'ODataResponseListTable',
      modelProperties: {
        odatacontext: {
          required: false,
          serializedName: 'odata\\.context',
          type: {
            name: 'String'
          }
        },
        value: {
          required: false,
          serializedName: 'value',
          type: {
            name: 'Sequence',
            element: {
                required: false,
                serializedName: 'TableElementType',
                type: {
                  name: 'Composite',
                  className: 'Table'
                }
            }
          }
        }
      }
    }
  };
};

module.exports = ODataResponseListTable;
