// powerbi-api/lib/autorest/models/oDataResponseListDatasource.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

var models = require('./index');

var util = require('util');

/**
 * @class
 * Initializes a new instance of the ODataResponseListDatasource class.
 * @constructor
 * Odata response wrapper for a Power BI datasource collection
 * @member {string} [odatacontext]
 * 
 * @member {array} [value] The datasource collection
 * 
 */
function ODataResponseListDatasource() {
}

/**
 * Defines the metadata of ODataResponseListDatasource
 *
 * @returns {object} metadata of ODataResponseListDatasource
 *
 */
ODataResponseListDatasource.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'ODataResponse[List[Datasource]]',
    type: {
      name: 'Composite',
      className: 'ODataResponseListDatasource',
      modelProperties: {
        odatacontext: {
          required: false,
          serializedName: 'odata\\.context',
          type: {
            name: 'String'
          }
        },
        value: {
          required: false,
          serializedName: 'value',
          type: {
            name: 'Sequence',
            element: {
                required: false,
                serializedName: 'DatasourceElementType',
                type: {
                  name: 'Composite',
                  className: 'Datasource'
                }
            }
          }
        }
      }
    }
  };
};

module.exports = ODataResponseListDatasource;
