// powerbi-api/lib/autorest/models/report.js
/*
 * Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */

'use strict';

/**
 * @class
 * Initializes a new instance of the Report class.
 * @constructor
 * A Power BI Report
 * @member {string} [id] The report id
 * 
 * @member {string} [name] The report name
 * 
 * @member {string} [webUrl] The report web url
 * 
 * @member {string} [embedUrl] The report embed url
 * 
 */
function Report() {
}

/**
 * Defines the metadata of Report
 *
 * @returns {object} metadata of Report
 *
 */
Report.prototype.mapper = function () {
  return {
    required: false,
    serializedName: 'Report',
    type: {
      name: 'Composite',
      className: 'Report',
      modelProperties: {
        id: {
          required: false,
          serializedName: 'id',
          type: {
            name: 'String'
          }
        },
        name: {
          required: false,
          serializedName: 'name',
          type: {
            name: 'String'
          }
        },
        webUrl: {
          required: false,
          serializedName: 'webUrl',
          type: {
            name: 'String'
          }
        },
        embedUrl: {
          required: false,
          serializedName: 'embedUrl',
          type: {
            name: 'String'
          }
        }
      }
    }
  };
};

module.exports = Report;
